File: packages/whisker-toggle/ios/Sources/WhiskerToggle/ToggleView.swift
import UIKit
import WhiskerModule

/// Minimal native control used as the first RFC0004 iOS element.
public final class ToggleView: WhiskerUI<UISwitch> {
    private var applyingProperty = false

    @objc public override func createView() -> UISwitch {
        let control = UISwitch()
        control.addTarget(self, action: #selector(valueChanged(_:)), for: .valueChanged)
        return control
    }

    public func setChecked(_ value: Bool) {
        applyingProperty = true
        self.view().setOn(value, animated: false)
        applyingProperty = false
    }

    public func setDisabled(_ value: Bool) {
        self.view().isEnabled = !value
    }

    @objc private func valueChanged(_ sender: UISwitch) {
        guard !applyingProperty else { return }
        WhiskerCustomEvent.dispatch(
            from: self,
            name: "change",
            params: ["checked": sender.isOn]
        )
    }
}

File: packages/whisker-toggle/android/src/main/kotlin/rs/whisker/toggle/ToggleView.kt
package rs.whisker.toggle

import android.content.Context
import android.view.View
import android.widget.Switch
import rs.whisker.runtime.WhiskerContext
import rs.whisker.runtime.WhiskerCustomEvent
import rs.whisker.runtime.WhiskerUI

/** Minimal native control used as the first RFC0004 Android element. */
public class ToggleView(context: WhiskerContext) : WhiskerUI<View>(context) {
    private var control: Switch? = null
    private var applyingProperty: Boolean = false

    override fun createView(context: Context): View = Switch(context).also { toggle ->
        control = toggle
        toggle.setOnCheckedChangeListener { _, checked ->
            if (!applyingProperty) {
                WhiskerCustomEvent.dispatch(
                    ui = this,
                    name = "change",
                    params = mapOf("checked" to checked),
                )
            }
        }
    }

    public fun setChecked(value: Boolean) {
        applyingProperty = true
        control?.isChecked = value
        applyingProperty = false
    }

    public fun setDisabled(value: Boolean) {
        control?.isEnabled = !value
    }
}
